Modularize distance and area formatting

m-kuhn opened the pull request:
All testing done so far is to validate, that it compiles :)

m-kuhn committed:
Modularize distance and area formatting

nyalldawson commented:
Looks good! Do you think scaled distance/... would belong better in QgsUnitTypes? (I do). 

I think there's a lot more code which needs updating when a new distance unit is added too...

nyalldawson commented on src/core/qgsunittypes.h:
```diff
@@ -49,6 +49,8 @@ class CORE_EXPORT QgsUnitTypes
       DistanceYards, //!< Imperial yards
       DistanceMiles, //!< Terrestrial miles
       DistanceDegrees, //!< Degrees, for planar geographic CRS distance measurements
+      DistanceCentimeters, //! Centimeters
```
Missing < on the docstrings

m-kuhn commented:
> Looks good! Do you think scaled distance/... would belong better in QgsUnitTypes? (I do).

Why not... I think it belongs next to `formatXyz`, `convertXyzMeasurement`, so they should probably be moved too.

nyalldawson commented:
Agreed. I just don't think most people would look in QgsDistanceArea for these functions.

Off topic: the name of QgsDistanceArea really bugs me. Shouldn't it be QgsDistanceAreaCalculator or QgsDistanceAreaMeasurer instead?

m-kuhn commented:
True, `QgsDistanceArea` implies storage of distances and areas. What I also don't like very much is that it lists the two terms "distance" and "area" and will need to be renamed when we get "volumes" 📦 . So maybe `QgsMeasurementCalculator` would be an alternative?

m-kuhn committed:
[travis] Fold segfault messages

nyalldawson commented:
Can't help but think this PR has got a little side tracked ;)

m-kuhn commented:
@nyalldawson should be back on track now. Anything you'd like to add?

nyalldawson commented:
All looks good to me - there's just some tests in test_qgsunittypes.py which need to be updated to add the new units. Just scroll through the existing tests and add cm/mm following the existing test setup.

nyalldawson commented on src/core/qgsunittypes.cpp:
```diff
@@ -864,6 +979,12 @@ QgsUnitTypes::AreaUnit QgsUnitTypes::distanceToAreaUnit( DistanceUnit distanceUn
     case DistanceKilometers:
       return AreaSquareKilometers;
 
+    case DistanceCentimeters:
```
This isn't right... I think you'll need to unfortunately add AreaSquareMillimeters/AreaSquareCentimeters to correctly handle this.

m-kuhn committed:
Unit tests for mm/cm units

m-kuhn committed:
[feature] Add support for square centimeters / square millimeter units

Note for documentation writers, centimers / millimeters support has been
added as well in a previous commit.

m-kuhn commented:
@nyalldawson all concerns should be addressed now.
Merge?

agiudiceandrea commented on src/core/qgis.h:
```diff
@@ -233,6 +233,17 @@ inline double qgsRound( double x )
+ */
+inline double qgsRound( double number, double places )
+{
+  int scaleFactor = pow( 10, places );
```
@m-kuhn, I think there is a flaw in qgsRound() that leads to incorrect rounding and/or overflow in some circumstances.
Could you please have a look at http://osgeo-org.1560.x6.nabble.com/QGIS-Developer-On-precision-in-qgsRound-3-3-0-master-td5377378.html and https://github.com/qgis/QGIS/pull/7836#issuecomment-419718554
